src/clang-c-frontend/clang_c_convert.cpp

```diff
@@ -640,6 +640,22 @@ bool clang_c_convertert::get_function(
       return false;
   }
 
+  // per https://eel.is/c++draft/dcl.spec.auto#general-14 return types of template functions are
+  // only deduced when they are instantiated (i.e. used).
+  // We skip those functions as they are unused anyway and the rest of esbmc can't handle undeduced types.
+  if (fd.getReturnType()->isUndeducedType() && !fd.isUsed(true))
```

Comment:
If being called through a function pointer, it should be used, because we have to take the address of it.
I can change this to just `ìf(fd.getReturnType()->isUndeducedType())` if you want.

Reply:
In the end, I included the `!fd.isUsed(true)` conjunct just in case there would ever be an undeduced type that **is** used. If we would ever hit such a case, we would crash in https://github.com/esbmc/esbmc/blob/bbd0ed41a74845b0f89b7152c1a6859d260be10a/src/clang-c-frontend/clang_c_convert.cpp#L1253
instead of just skipping the function and having surprising problems later on.

Reply:
Can we have undeduced types in c code? Otherwise, we can move this check to `get_method` instead.

Reply:
I moved the part that checks whether the function is used to an assert.

Reply:
> please remove the isUsed call, as it is not reliable.

In all my testing I did not find any case where `isUsed` was not reliable.
It's also in an assert so only devs will be affected should there ever be a false result.

Reply:
Your tests do not cover all possible C++ programs in existence. 

We release esbmc with RelWithDebInfo, so asserts are enabled for users too.

Remove the isUsed assertion.